Subclassing round 10

bilelmoussaoui opened the pull request:
Adds subclassing support of the following types
- GtkTreeModel
- GtkTreeSortable

Part of #56

sdroege commented on gtk4/src/subclass/font_chooser.rs:
```diff
@@ -0,0 +1,196 @@
+    fn get_font_family(&self, font_chooser: &Self::Type) -> Option<FontFamily>;
+    fn get_font_face(&self, font_chooser: &Self::Type) -> Option<FontFace>;
+    fn get_font_size(&self, font_chooser: &Self::Type) -> i32;
+    // fn set_filter_func(&self, font_chooser: &Self::Type) -> Option<Box<dyn Fn(&FontFamily, &FontFace) -> bool + 'static>>;
```
This is the wrong way around. The caller would pass a filter function in here.

You'd need to define a new type for that which wraps the function pointer, user_data and destroy notify, which can be called from safe Rust, etc.

I think we have such a case elsewhere in GTK already. The foreach function of GtkContainer or so

sdroege replied:
Also this function is not optional. You have to add bindings for it.

sdroege commented on gtk4/src/subclass/font_chooser.rs:
```diff
@@ -0,0 +1,196 @@
+    let ret = imp.get_font_family(wrap.unsafe_cast_ref());
+    if let Some(font_family) = ret {
+        let font_family = font_family.to_glib_full();
+        wrap.set_qdata(*FONT_CHOOSER_GET_FONT_FAMILY_QUARK, font_family);
```
You'd leak this pointer here. That's probably not intended

sdroege replied:
Also same below

bilelmoussaoui replied:
> You'd leak this pointer here. That's probably not intended

But the fonction annotates the `FontFamily`/`FontFace` as `return none`, or I misunderstood something? :)

sdroege replied:
You store a raw pointer in the qdata. That pointer is never freed. You would have to store some struct around a raw pointer in the qdata that implements `Drop` in a way that the pointer is freed.

bilelmoussaoui replied:
I have did something, but not very sure if what I wrote makes sense or not :)

sdroege replied:
Looks good, make it `ptr::NonNull` instead of a plain raw pointer and it's perfect :)

sdroege commented on gtk4/src/subclass/tree_model.rs:
```diff
@@ -0,0 +1,503 @@
+    fn get_iter(&self, tree_model: &Self::Type, path: &TreePath) -> Option<TreeIter>;
+    fn get_path(&self, tree_model: &Self::Type, iter: &TreeIter) -> TreePath;
+    fn get_value(&self, tree_model: &Self::Type, iter: &TreeIter, index: i32) -> Value;
+    fn iter_next(&self, tree_model: &Self::Type) -> Option<TreeIter>;
```
Can we actually implement all these functions? Do we have ways to create a tree iter out of thin air?

bilelmoussaoui replied:
We don't actually have a way to create a TreeIter currently no, but it's just a sturct around an int & 3 random user_data pointers; I will have to look at how other bindings handle this (if they do)

bilelmoussaoui replied:
So none of the bindings provides a way to create a TreeIter but as we made the whole trait unsafe, I guess that's okay as is.

sdroege replied:
I'd rather not have this trait for now then. It clearly shows the it's not usable yet and not enough thought was put into how to make it work in Rust. Having the trait around in an unusable form will just confuse users that try to use it.

Maybe move this into a separate WIP PR?

bilelmoussaoui replied:
Sure, I can move the FontChooser out of this PR as the rest is dependant on TreeModel. It could definitely use some better approach to make it usable from Rust but requires someone who is actually familiar with the Tree* stuff in GTK :) 
Also I couldn't find that many apps that has custom TreeModels, most of them are written in C like sysprof.

sdroege commented on gtk4/src/subclass/tree_model.rs:
```diff
@@ -0,0 +1,503 @@
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        &iter,
+    )
+    .to_glib_full() as *mut _
```
This misses an annotation in the C code btw

bilelmoussaoui replied:
alot of thing are not annotated in TreeModel*  :/ I will try to bring that up for gtk4.2 and hopefully fix them. The thing is I'm afraid of that nobody touched for a decade code.

sdroege commented on gtk4/src/subclass/tree_model.rs:
```diff
@@ -0,0 +1,503 @@
+        &iter,
+        index,
+    );
+    *valueptr = *ret.to_glib_none().0;
```
This is wrong. You'd need to `mem::forget()` the value afterwards or use `ManuallyDrop` (better), otherwise the content of the value will be freed twice

Also the C annotations are completely wrong. This should be `out caller-allocates` AFAIU.

sdroege commented on gtk4/src/subclass/tree_model.rs:
```diff
@@ -0,0 +1,503 @@
+    ) -> Option<TreeIter>;
+    fn iter_parent(&self, tree_model: &Self::Type, child: &TreeIter) -> Option<TreeIter>;
+
+    fn ref_node(&self, tree_model: &Self::Type, iter: &TreeIter) {
```
Can we safely (or at all) implement this and the unref function? How is that supposed to work?

bilelmoussaoui replied:
The API docs mentions that the functions handles a ref counter of the currently displayed nodes so that they are cached. Here's an example from gtk of how such methods looks like 
```c
static void
gtk_tree_model_ref_count_ref_node (GtkTreeModel *model,
                                   GtkTreeIter  *iter)
{
  NodeInfo *info;
  GtkTreeModelRefCount *ref_model = GTK_TREE_MODEL_REF_COUNT (model);

  info = g_hash_table_lookup (ref_model->priv->node_hash, iter->user_data);
  if (!info)
    {
      info = node_info_new ();

      g_hash_table_insert (ref_model->priv->node_hash, iter->user_data, info);
    }

  info->ref_count++;
}

static void
gtk_tree_model_ref_count_unref_node (GtkTreeModel *model,
                                     GtkTreeIter  *iter)
{
  NodeInfo *info;
  GtkTreeModelRefCount *ref_model = GTK_TREE_MODEL_REF_COUNT (model);

  info = g_hash_table_lookup (ref_model->priv->node_hash, iter->user_data);
  g_assert (info != NULL);
  g_assert (info->ref_count > 0);

  info->ref_count--;
}
```
But I'm not sure where that ref counting is used, as in where the caching happens. I guess we can mark them as unsafe and leave figuring out how this part of the api is supposed to work to someone that actually uses GtkTreeModel

sdroege replied:
That would mean making the trait unsafe, not the function.

sdroege commented on gtk4/src/subclass/tree_sortable.rs:
```diff
@@ -0,0 +1,176 @@
+/*
+TODO: figure out
+- set_sort_func
+- sset_default_sort_func
```
They're all mandatory so you'd have to implement them before it can be merged.

sdroege commented on gtk4/src/subclass/tree_sortable.rs:
```diff
@@ -0,0 +1,176 @@
+                as *mut ffi::GtkTreeSortableIface;
+            assert!(!iface.is_null());
+
+            if let Some(f) = (*iface).sort_column_changed.as_ref() {
```
There are no default impls for anything here

bilelmoussaoui commented on gtk4/src/subclass/font_chooser.rs:
```diff
@@ -0,0 +1,231 @@
+use pango::{FontFace, FontFamily, FontMap};
+
+#[derive(Debug)]
+pub struct FilterCallback {
```
I wonder if we should provide some way to convert a rust callback to a FilterCallback or do we expect the user to write unsafe code in order to implement the FontChooser?

sdroege replied:
That safe function already exists: https://gtk-rs.org/gtk4-rs/gtk4/prelude/trait.FontChooserExt.html#tymethod.set_filter_func

You get this callback passed via C and in the implementation side of the font chooser you only need to pass it around and call it.

sdroege replied:
I.e. you need to also provide a way to safely call this, see https://github.com/gtk-rs/gtk-rs/blob/c58fab4fec1254244f598c4eac7f9310184c5996/gtk/src/subclass/container.rs#L258-L266

And you'll have to understand how this works memory management wise. Does the subclass own it and has to free it? Or the base class? You will probably have to implement `Drop` on this, which then would pass the `user_data` into the `destroy_notify` (if the `destroy_notify` is not `NULL`).

sdroege commented on gtk4/src/subclass/tree_model_filter.rs:
```diff
@@ -0,0 +1,142 @@
+
+use crate::{TreeIter, TreeModel, TreeModelFilter};
+
+pub trait TreeModelFilterImpl: TreeModelFilterImplExt + ObjectImpl {
```
```suggestion
pub trait TreeModelFilterImpl: TreeModelFilterImplExt + TreeModelImpl + TreeDragSourceImpl + ObjectImpl {
```

If I read the docs correctly

sdroege replied:
Ah nevermind

sdroege commented on gtk4/src/subclass/tree_sortable.rs:
```diff
@@ -0,0 +1,224 @@
+    destroy_ptr: glib::ffi::GDestroyNotify,
+}
+
+pub trait TreeSortableImpl: ObjectImpl {
```
```suggestion
pub trait TreeSortableImpl: TreeModelImpl + ObjectImpl {
```

Though. See the `Prerequisites` in the docs.

sdroege commented on gtk4/src/subclass/tree_sortable.rs:
```diff
@@ -0,0 +1,224 @@
+pub struct TreeIterCompareCallback {
+    compare_func: ffi::GtkTreeIterCompareFunc,
+    user_data: glib::ffi::gpointer,
+    destroy_ptr: glib::ffi::GDestroyNotify,
```
Same question about memory management here

sdroege replied:
And needs to be callable

bilelmoussaoui commented:
I have rebased this one and updated per the latest subclassing changes, just in case someone wants to have a look at it someday :) it should make things simpler to pick up.

bilelmoussaoui commented:
Closing this one as I won't be finishing it. People are free to take over from that branch or just switch to using Gtk*View widgets instead
